introduce performance profiling doc (#167)
operator-framework/olm-docs: content/en/docs/Tasks/performance-profiling.md

---

> title: "Performance Profiling Metrics"

Reviewer: I can imagine readers finding this page while trying to understand the performance of their own operator versus the OLM operators. Making the title more specific might save them some confusion or at least a click.

weight: 10
description: >

>   The goal of this document is to familiarize you with the steps to enable and review OLM's performance profiling metrics.

Reviewer: ```suggestion
  The goal of this document is to familiarize you with the steps to enable and review OLM's performance profiling instrumentation.
```

---

## Prerequisites 

- [go](https://golang.org/dl/)

## Background

> OLM utilizes the [pprof package](https://golang.org/pkg/net/http/pprof/) from the standard go library to expose performance profiles for the OLM Operator, the Catalog Operator, and Registry Servers. Due to the sensitive nature of this data, OLM must be configured to use TLS Certificates before performance profiling can be enabled.

Reviewer: > and Registry Servers

Not registry-server today.

> must be configured to use TLS Certificates before performance profiling can be enabled

For clarity, since the profiling endpoints are always enabled now, this could probably be combined with the following paragraph and say something like "due to the sensitive nature of profiling data, the profiling endpoints will reject any clients that do not present a verifiable certificate. Both operators must be configured with a serving certificate and a client CA bundle in order to access the profiling endpoints".

Requests against the performance profiling endpoint will be rejected unless the client certificate is validated by OLM. Unfortunately, Kubernetes does not provide a native way to prevent pods on cluster from iterating over the list of available ports and retrieving the data exposed. Without authenticating the requests, OLM could leak customer usage statistics on multitenant clusters.

> This document will dive into the steps to [enable olm performance profiling](enable-performance-profiling) and retrieving pprof data from each component.

Reviewer 2: It looks like a broken link to the next header section.

```suggestion
This document will dive into the steps to [enable olm performance profiling](#enabling-performance-profiling) and retrieving pprof data from each component.
```

## Enabling Performance Profiling

### Creating a Certificate

> A valid server certiciate must be created for each component before Performance Profiling can be enabled. If you are unfamiliar with certificate generation, I recomend using the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/manmaster/man1/openssl-req.html) documentation.

Reviewer 2: ```suggestion
A valid server certificate must be created for each component before the Performance Profiling functionality can be enabled. If you are unfamiliar with certificate generation, it's recommend to use the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/man1.0.2/man1/openssl-req.html) documentation.
```

Once you have generated a private and public key, this data should be stored in a `TLS Secret`:

```bash

$ export TLS_SECRET=my-tls-secret
$ export PRIVATE_KEY_FILENAME=private.key # Replace with the name of the file that contains the private key you generated.
$ export PUBLIC_KEY_FILENAME=certificate.crt # Replace with the name of the file that contains the public key you generated.

> $ cat << EOF | kubectl apply -f -

Reviewer: Users [can also do](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets):

```
kubectl -n my-namespace create secret tls my-name --cert=certfile --key=keyfile
```

apiVersion: v1
kind: Secret
metadata:
  name: $TLS_SECRET
  namespace: olm
type: kubernetes.io/tls
data:
  tls.key: $(base64 -w0 $PRIVATE_KEY_FILENAME)
  tls.crt: $(base64 -w0 $PUBLIC_KEY_FILENAME)
EOF
```

> ### Updating OLM to Use the TLS Secret

Reviewer: I know the chart is sort of an internal detail of the release process, but since the audience for this doc is skewed towards developers, maybe it'd be worth putting some commented-out-with-explanation lines into the values.yaml and mentioning them here as well?

Patch the OLM Deployment's pod template to use the generated TLS secret:

- Defining a volume and volumeMount

> - Adding the `client-ca`, `tls-key` and `tls-crt` arguments

Reviewer: ```suggestion
- Adding the `client-ca`, `tls-key` and `tls-cert` arguments
```

- Replacing all mentions of port `8080` with `8443`

> - Updating the `livenessProbe` and `readinessProbe` to use HTTPS as the scheme.

Reviewer 2: ```suggestion
- Updating the `livenessProbe` and `readinessProbe` to use HTTPS as the scheme
```

> This can by generating patching an existing OLM deployment:

Reviewer 2: ```suggestion
This can by generated by patching an existing OLM deployment:
```

    {"op": "replace", "path": "/spec/template/spec/containers/0/ports/0", "value":{"containerPort": 8443}},
    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/port", "value":8443},
    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/port", "value":8443},

    # Update livenessProbe and readinessProbe to use HTTPS
    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/scheme", "value":"HTTPS"},
    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/scheme", "value":"HTTPS"},
]'
deployment.apps/olm-operator patched

```

## Accessing PPROF Data

You will need to be able to access OLM port, for dev purposes the following commands may prove useful:

```bash
# Set Deployment name to olm-operator or catalog-operator
$ export DEPLOYMENT_NAME=olm-operator
$ kubectl port-forward deployment/$DEPLOYMENT_NAME 8443:8443 -n olm
```

You can then curl the OLM `/debug/pprof` endpoint to retrieve default pprof profiles like so:

```bash
$ curl https://localhost:8443/debug/pprof/heap --cert certificate.crt --key private.key  --insecure -o olm-heap

> $ go tool pprof --top olm-heap

Reviewer: This could be a good place to drop a link to something like https://blog.golang.org/pprof to get readers started with how to make sense of the profiles.

---
title: "Continuous Profiling"
weight: 10
description: >
  The goal of this document is to familiarize you with OLM's stance on performance profiling.
---

## Prerequisites 

- [go](https://golang.org/dl/)

## Background

OLM utilizes the [pprof package](https://golang.org/pkg/net/http/pprof/) from the standard go library to expose performance profiles for the OLM Operator, the Catalog Operator, and Registry Servers. Due to the sensitive nature of this data, client requests against the pprof endpoint are rejected unless they are made with the certificate data kept in the `pprof-cert secret` in the `olm namespace`. 

> Kubernetes does not provide a native way to prevent pods on cluster from iterating over the list of available ports and retrieving the data exposed. Without authetnicating the requests, OLM could leak customer usage statistics on multitenant clusters. If the aforementioned secret does not exist the pprof data will not be accessable.

Reviewer 2: Just a quick driveby as I see this is still in a draft:

authetnicating -> authenticating

This document will dive into the steps to [enable olm performance profiling](enable-performance-profiling) and retrieving pprof data from each component.

## Enabling Performance Profiling

### Creating a Certificate

A valid server certiciate must be created for each component before Performance Profiling can be enabled. If you are unfamiliar with certificate generation, I recomend using the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/manmaster/man1/openssl-req.html) documentation.

Once you have generated a private and public key, this data should be stored in a `TLS Secret`:

```bash
$ export PRIVATE_KEY_FILENAME=private.key # Replace with the name of the file that contains the private key you generated.
$ export PUBLIC_KEY_FILENAME=certificate.key # Replace with the name of the file that contains the public key you generated.

$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:

>   name: olm-serving-secret

Author: ```suggestion
  name: olm-serving-cert
```

    {"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value":"'$CERT_PATH'/tls.crt"},
    {"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value":"--tls-key"},
    {"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value":"'$CERT_PATH'/tls.key"},
    {"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value":"--tls-cert"},
    {"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value":"'$CERT_PATH'/tls.crt"},
    
    # Replace port 8080 with 8443
    {"op": "replace", "path": "/spec/template/spec/containers/0/ports/0", "value":{"containerPort": 8443}},
    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/port", "value":8443},
    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/port", "value":8443},

    # Update livenessProbe and readinessProbe to use HTTPS
    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/scheme", "value":"HTTPS"},
    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/scheme", "value":"HTTPS"},
]'

> deployment.apps/olm-operator patched

Author: Alternatively, we could generate the manifests that include these changes as part of the OLM release.
